examples/smart-contract-upgrade/Contract_Version1/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+#[receive(contract = "smart_contract_upgrade", name = "view", return_value = "State")]
+fn contract_view<'b, S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &'b impl HasHost<State, StateApiType = S>,
+) -> ReceiveResult<&'b State> {
+    Ok(host.state())
+}
+
+/// Upgrade this smart contract instance to a new module and call optionally a
+/// migration function after the upgrade.
+///
+/// It rejects if:
+/// - Sender is not the admin of the contract instance.
+/// - It fails to parse the parameter.
+/// - If the ugrade fails.
+/// - If the migration invoke fails.
+///
+/// This function is marked as `low_level`. This is **necessary** since the
+/// high-level mutable functions store the state of the contract at the end of
+/// execution. This conflicts with migration since the shape of the state
+/// **might** be changed by the migration function. If the state is then written
+/// by this function it would overwrite the state stored by the migration
+/// function.
+#[receive(
+    contract = "smart_contract_upgrade",
+    name = "upgrade",
+    parameter = "UpgradeParams",
+    error = "CostumContractError",
+    low_level
+)]
+fn contract_upgrade<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<S>,
+) -> ContractResult<()> {
+    // Read the top-level contract state.
+    let state: State = host.state().read_root()?;
+
+    // Check that only the admin is authorized to upgrade the smart contract.
+    ensure!(ctx.sender().matches_account(&state.admin), CostumContractError::Unauthorized);
+    // Parse the parameter.
+    let params: UpgradeParams = ctx.parameter_cursor().get()?;
+    // Trigger the upgrade.
+    host.upgrade(params.module)?;
+    // Call the migration function if provided.
+    if let Some((func, parameters)) = params.migrate {
+        host.invoke_contract_raw(
+            &ctx.self_address(),
+            parameters.as_parameter(),
+            func.as_entrypoint_name(),
+            Amount::zero(),
+        )?;
+    }
+    Ok(())
+}
+
+/// Transfer the admin address to a new admin address.
+///
+/// It rejects if:
+/// - Sender is not the current admin of the contract instance.
+/// - It fails to parse the parameter.
+#[receive(
+    contract = "smart_contract_upgrade",
+    name = "updateAdmin",
+    parameter = "AccountAddress",
+    error = "CostumContractError",
+    mutable
+)]
+fn contract_update_admin<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State, StateApiType = S>,
+) -> ContractResult<()> {
+    // Check that only the current admin is authorized to update the admin address.
+    ensure!(ctx.sender().matches_account(&host.state().admin), CostumContractError::Unauthorized);
+
+    // Parse the parameter.
+    let new_admin = ctx.parameter_cursor().get()?;
+
+    // Update the admin variable.
+    host.state_mut().admin = new_admin;
+
+    Ok(())
+}
+
+// TODO: Add tests once the new integration testing library is released.
```

Comment:
Maybe make a small issue instead of this TODO, so that it is harder to forget :)

Reply:
https://github.com/Concordium/concordium-rust-smart-contracts/issues/264

Reply:
Thank you.